report/3.Theory/Lexer.tex

```diff
@@ -0,0 +1,72 @@
+
+The output of the scanner will be a stream of tokens.
+A token simply corresponds to indivual symbols in a given programming language.
+The scanner will attempt to match every lexeme to a token.
+Sometimes several lexemes can be matched by the same token.
+In the case of identifiers, two different variables might be named \textit{Employee} and \textit{Customer}, but they would both be matched to the identifier token.
+Some tokens can only match a single, very specific, lexeme.
+An example of this would be parentheses, where the left parentheses character, \textbf{(}, would only match the left parentheses token, and likewise for the right parentheses character.
+
+The point here is to illustrate the importance of knowing when a parentheses starts or ends, compared to just knowing that there is an identifier, but not which identifier
+
+
+Creating a lexer can be done in a number of ways.
+%Possibly insert a section here about Regular Expressions? Jones)
+Below the different options will be explored.
+\subsection{Lexer Generators}
+The primary advantage for all Lexer generators, is that they allow for easy modification of the language, without having to  rewrite the lexer yourself, as it is usually as simple as adding the rule to the CFG, and running the generator.
+Below we will briefly describe ANTLR and Flex, two of the most well-known and used Lexer/Parser generators.
+
+\textbf{ANTLR}\\
+ANTLR is a scanner/parser generator, which when given a context-free grammar, expressed using Extended Backus-Naur form (EBNF), can generate lexers, parsers and lexer-parsers\cite{ANTLR}.
+ANTLR can output to Java, C\#, Python2 and Python3, Javascript, Go, C++ and Swift\cite{ANTLRDocsTargets}.
+One of the primary advantages of ANTLR is that it uses a single notation for both lexers and parsers\cite{ANTLRDocs}.
+This means that after the lexer has been implemented, it would be extremely easy to implement the parser afterwards, as it would be as simple as specifying that a parser needs to be generated.
+As one of the primary design decisions of is simplicity over complexity, as well as the fact that ANTLR is often used in academia, it would be a good choice for a lexer/parser generator.
+ANTLR has official plugins for both Intellij, NetBeans and Eclipse, which improves the user experience, by supporting syntax highlighting, formatting and error detection.
+%Maybe add example syntax?
+
+\textbf{Lex}\\
+At the 4th semester, the Crafting a compiler uses Lex as an example of a Lexer generator\cite{CraftingACompiler}.
+Lex is by many deemed the standard lexer generator, and it comes default on many unix systems.
+Lex outputs to C, however there are many different forks which allow for output to different languages.
+A Lex file is comprised of 3 distinct sections\cite{Lex}.
+\begin{itemize}
+	\item \textbf{Definition}, in which will import header files and allows for definition of macros. 
```

Comment:
"in which will import" <- in which what will import?
 are you missing an identifier?